core/src/main/python/wlsdeploy/tool/discover/topology_discoverer.py

```diff
@@ -425,6 +427,25 @@ def _get_jmx(self):
         _logger.exiting(class_name=_class_name, method_name=_method_name)
         return model_top_folder_name, result
 
+    def _get_jpa(self):
+        """
+        Discover the domain level JPA configuration attributes.
+        :return: model name for JPA:dictionary containing the discovered JPA attributes
+        """
+        _method_name = '_get_jpa'
+        _logger.entering(class_name=_class_name, method_name=_method_name)
+        model_top_folder_name = model_constants.JPA
+        result = OrderedDict()
+        location = LocationContext(self._base_location)
+        location.append_location(model_top_folder_name)
+        name = self._find_singleton_name_in_folder(location)
+        if name is not None:
+            _logger.info('WLSDPLY-06644', class_name=_class_name, method_name=_method_name)
+            location.add_name_token(self._alias_helper.get_name_token(location), name)
+            self._populate_model_parameters(result, location)
+        _logger.exiting(class_name=_class_name, method_name=_method_name)
+        return model_top_folder_name, result
+
```

Comment:
This is the fifth almost-identical method for discovering a single-unpredictable MBean in this class (sixth if including sub-folders). The only variation is the folder name and error code. Should we create a more general method (preferably in the parent class Discoverer), passing the folder name and using a more general error message, such as "Adding {0}"? Not suggesting a refactor of all discoverer classes, just this one for now.

Reply:
I originally wanted to do something a little more complex in my refactor, but will go ahead and start your suggestion. In future, it can be refactored differently.